### PR TITLE
Add `StdoutConfig` and `StderrConfig` to steps.

### LINKED ISSUE #2925: Make it possible to extract results from a container's stdout
### Feature request

I'd like to be able to create a Task which provides results with come from the output of a container, without having to capture that input and parse it in my Task.

Initially we could do this only when output is json formatted maybe?

For example do something like this:

```yaml
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: boskos-acquire
spec:
    steps:
    - name: boskosctl
      image: gcr.io/k8s-staging-boskos/boskosctl@sha256:a7fc984732c5dd0b4e0fe0a92e2730fa4b6bddecd0f6f6c7c6b5501abe4ab105 
      args:
      - "acquire"
      - "--server-url=http://boskos.test-pods.svc.cluster.local"
      - "--owner-name=christie-test-boskos"
      - "--type=gke-project"
      - "--state=free"
      - "--target-state=busy"
     results:
     - name: project-id
       description: The name of the project acquired
       fromStep:
       - name: boskosctl
          jsonPath: .name
```

The `fromStep` above would express the jsonpath to the desired value to use as the result, which is expected to be in the stdout of the command.

### Use case

I'm working on a boskos Task as part of the catalog test infrastructure. This is what I'm playing around with now:

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  generateName: boskosctl-test-
spec:
  taskSpec:
    steps:
    - image: gcr.io/k8s-staging-boskos/boskosctl@sha256:a7fc984732c5dd0b4e0fe0a92e2730fa4b6bddecd0f6f6c7c6b5501abe4ab105 
      args:
      - "acquire"
      - "--server-url=http://boskos.test-pods.svc.cluster.local"
      - "--owner-name=christie-test-boskos"
      - "--type=gke-project"
      - "--state=free"
      - "--target-state=busy"
```

This lets me acquire a boskos project which is great, but the only way to tell what project I've acquired is to look at the container logs, e.g.:

```
[build-step-unnamed-1] {"type":"gke-project","name":"tekton-prow-6","state":"busy","owner":"christie-test-boskos","lastupdate":"2020-07-09T20:02:41.730506184Z","userdata":{}}
```

So even tho the boskos folks provided this sweet `boskosctl` image with a reasonable `cmd` default, im gonna end up having to override the entrypoint and write a script (and hope that the image has bash in it!)

### PR DESCRIPTION
**Pull request by chhsia0**
# Changes

This PR implements Option 1 of [TEP-0011](https://github.com/tektoncd/community/blob/master/teps/0011-redirecting-step-output-streams.md).

Closes #2925.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

**EXPERIMENTAL**: Users can specify the `TEKTON_STDOUT_FILE` and `TEKTON_STDERR_FILE` environment variables in steps to capture steps' stdout and stderr to local files. This feature can be used to capture stdout and stderr into task results.

### CODE REVIEW

File: cmd/entrypoint/runner.go [hunk 1]
```diff
@@ -34,8 +37,51 @@ func (rr *realRunner) Run(args ...string) error {
...
+		if stdoutFile, err = os.Create(rr.stdoutFile); err != nil {
+			return err
+		}
+		defer stdoutFile.Close()
+		// We use os.Pipe in asyncWriter to copy stdout instead of cmd.StdoutPipe or providing an
+		// io.Writer directly because otherwise Go would wait for the underlying fd to be closed by the
+		// child process before returning from cmd.Wait even if the process is no longer running. This
+		// would cause a deadlock if the child spawns a long running descendant process before exiting.
+		stopCh := make(chan struct{}, 1)
+		defer close(stopCh)
```
Comment (vdemeester) [inline cmd/entrypoint/runner.go:54]: I think we shouldn't need to create 2 `stopCh` (for `stdout` and `stderr`), on is enough :upside_down_face:

Comment (chhsia0) [reply cmd/entrypoint/runner.go:54]: Oh you're right. Thanks!

**Comment by chhsia0** [comment]
@vdemeester I'll create a TEP for passing the params :) Right now in its current state user would use env vars as a channel transparent to the Tekton controller. Sounds scary huh? ;)

**Comment by vdemeester** [comment]
> @vdemeester I'll create a TEP for passing the params :) Right now in its current state user would use env vars as a channel transparent to the Tekton controller. Sounds scary huh? ;)

:grin: nothing is scary :stuck_out_tongue_closed_eyes: and as you said during the API WG, it allows you to more easily experiment with it without really commiting to it (aka see if it's useful and all).

**Review by sbwsg — COMMENTED** [review]
I've done a quick first pass read-through but need to spend much more time wrapping my head around the change.  There's quite a lot going on in `cmd/entrypoint/io.go` and I have to confess I'm sort of lost as to the necessity for all the go routines.  Given that we have unit tests for the content in `io.go`, I wonder if it would make sense to put it into its own package along with its tests and export its methods publicly. We could then document its purpose and approach in a bit more detail using fuller docstrings and possibly a `doc.go` file?

Also, unless there's been an intentional package add or upgrade I suggest dropping the changes in go.mod and go.sum in favor of upgrading packages in a separate PR.

**Comment by vdemeester** [comment]
@chhsia0 any update on this ? Do you need help for this (one of us could carry the PR toward merging) ?

**Comment by bobcatfish** [comment]
highly relevant: https://github.com/kubernetes/kubernetes/issues/94892

**Comment by bobcatfish** [comment]
@chhsia0 I'm thinking maybe it makes sense at this point to close this PR for now, unless you're planning to pick it up again in the near future. We can always re-open.

**Comment by vdemeester** [comment]
I think we can close, and I plan to carry this one soon 😇
/close

**Pull request closed without merging**